pkg/router/route.go

```diff
@@ -0,0 +1,161 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package router
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+)
+
+import (
+	"github.com/emirpasic/gods/trees/avltree"
+	"github.com/pkg/errors"
+)
+
+import (
+	"github.com/dubbogo/dubbo-go-proxy/pkg/common/constant"
+	"github.com/dubbogo/dubbo-go-proxy/pkg/config"
+)
+
+// Node defines the single method of the router configured API
+type Node struct {
+	fullPath string
+	wildcard bool
+	methods  map[config.HTTPVerb]*config.Method
+}
+
+// Route defines the tree of router APIs
+type Route struct {
+	lock         sync.RWMutex
+	tree         *avltree.Tree
+	wildcardTree *avltree.Tree
+}
+
+// Put put a key val into the tree
+func (rt *Route) Put(fullPath string, method config.Method) error {
+	rt.lock.Lock()
+	defer rt.lock.Unlock()
+	fullPath = strings.ToLower(fullPath)
+	wildcard := containParam(fullPath)
+
+	if wildcardNode, found := rt.searchWildcard(fullPath); found {
+		return putMethod(wildcardNode, method)
+	}
+
+	node, ok := rt.tree.Get(fullPath)
+	if !ok {
+		ms := make(map[config.HTTPVerb]*config.Method)
+		rn := &Node{
+			fullPath: fullPath,
+			methods:  ms,
+			wildcard: wildcard,
+		}
+		rn.methods[method.HTTPVerb] = &method
+		if wildcard {
+			rt.wildcardTree.Put(fullPath, rn)
+		}
+		rt.tree.Put(fullPath, rn)
+		return nil
+	}
+	return putMethod(node.(*Node), method)
+}
+
+// UpdateMethod update the api method in the existing router node
+func (rt *Route) UpdateMethod(fullPath string, verb config.HTTPVerb, method config.Method) error {
+	rt.lock.Lock()
+	defer rt.lock.Unlock()
+	node, found := rt.findNode(fullPath)
+	if found {
+		if _, ok := node.methods[verb]; ok {
+			node.methods[verb] = &method
+		}
+	}
+	return nil
+}
+
+func (rt *Route) findNode(fullPath string) (*Node, bool) {
+	fullPath = strings.ToLower(fullPath)
+	var n interface{}
+	var found bool
+	if n, found = rt.searchWildcard(fullPath); !found {
+		n, found = rt.tree.Get(fullPath)
+	}
+	return n.(*Node), found
+}
+
+// FindMethod returns the api that meets the
+func (rt *Route) FindMethod(fullPath string, httpverb config.HTTPVerb) (*config.Method, bool) {
+	if n, found := rt.findNode(fullPath); found {
+		method, ok := n.methods[httpverb]
+		return method, ok
+	}
+	return nil, false
+}
+
+func (rt *Route) searchWildcard(fullPath string) (*Node, bool) {
+	wildcardPaths := rt.wildcardTree.Keys()
+	for _, p := range wildcardPaths {
+		if wildcardMatch(p.(string), fullPath) {
+			n, ok := rt.wildcardTree.Get(p)
+			return n.(*Node), ok
+		}
+	}
+	return nil, false
+}
+
+func putMethod(node *Node, method config.Method) error {
+	if _, ok := node.methods[method.HTTPVerb]; ok {
+		return errors.New(fmt.Sprintf("Method %s already exists in path %s", method.HTTPVerb, node.fullPath))
+	}
+	node.methods[method.HTTPVerb] = &method
```

Comment:
`putMethod` should be a method of `Node`? and add a lock for this mehod?

Reply:
done